File: ak_tactic/frontend/stage_env.py
# -*- coding: utf-8 -*-
"""**关卡静态**:构建规格要用、但**与战斗无关**的那 8 个量。

## 它替换掉什么

`build_spec` 现在从**活的 `BattleSimulator`** 上读这 8 项
（`spec.py:1062-1078`）：

```
fps  speed_scale  ranged_enemies  enemy_windup
cost_init  cost_max  cost_time  life
```

而它们在 `sim.py` 那边的来源是 `sim.py:418-489`——**全是派生**，
没有一项需要引擎跑起来：

| 项 | 源头 |
|---|---|
| `fps` | 构造参数（`sim.py:418`） |
| `speed_scale` | 构造参数 × `stage.options.move_multiplier`（419） |
| `ranged_enemies` | 构造参数（428） |
| `enemy_windup` | `max(0, 构造参数)`（440） |
| `cost_init` | `stage.options.initial_cost`（474） |
| `cost_max` | `stage.options.max_cost`（475） |
| `cost_time` | `stage.options.cost_increase_time` **÷** `cost_recovery_scale(...)`（476/479-481） |
| `life` | `global_lifepoint(...)` 优先，否则 `stage.options.max_life_point`（483/487-489） |

## ⚠ 两处容易漏的派生（它们正是"四星档"的来历）

* `cost_time` 要被 `cbuff_cost_recovery.scale` **除**（四星档常见 2 = 回复速度翻倍）。
  1-7 的四星档就是这条。**乘除弄反不会有任何判据报警**，只会让费用回得快/慢。
* `life` 要被关卡级的 `global_lifepoint` **覆写**（八关 EX 的四星档都改成 1，
  而关卡文件自己的 `maxLifePoint` 是 3、普通与四星**两份都是 3**）。
  不接这条，四星档就凭空多两条命。

## 判据边界（照抄 `blackboard.py` 那段）

这两条都只在**非 NORMAL 难度**下才与普通档不同，而当前流水线**恒为 NORMAL**
（`Verifier` 建模拟器时不传 `environment_difficulty`）⇒ 那是**沉默区**。

**所以本模块是"原样搬"，不是重写**：上面每一行的口径都与 `sim.py` 逐字对齐，
包括 `or 0.0` / `or 99.0` / `or 1.0` 这些"0 与 None 一视同仁"的写法
（改成 `is None` 判断会在 `max_cost=0` 这类取值上分叉，而现有计划覆盖不到）。
"""
from __future__ import annotations

from typing import Any

from .stage_mul import cost_recovery_scale, global_lifepoint

__all__ = ["FPS", "stage_env", "ENV_KEYS"]

#: 每秒帧数。⚠ **一份来源**：`battle/sim.py` 从这里 import（原来是 `sim.py:79`
#: 自己写的 30）。两处各写一个 30 的话，哪天要改就会只改一处。
FPS = 30

#: 本模块产出的键（规格里对应的字段名）。
ENV_KEYS = ("fps", "speed_scale", "ranged_enemies", "enemy_windup",
            "cost_init", "cost_max", "cost_time", "life")


def stage_env(stage: Any, *, environment_difficulty: str = "NORMAL",
              fps: int = FPS, speed_scale: float = 1.0,
              ranged_enemies: bool = True,
              enemy_windup: float = 0.5) -> dict[str, Any]:
    """`stage` ＋ 那几个构造参数 → 规格里那 8 个字段。

    后四个参数的默认值与 `BattleSimulator.__init__`（`sim.py:324-330`）逐字一致，
    这样"不传 switches"这条常见路径得到的就是同一个值。
    """
    # ---- 构造参数那四个（`sim.py:418-440`）----
    out: dict[str, Any] = {
        "fps": int(fps),
        # `move_multiplier` 是关卡级的移速档（`stage.options`）
        "speed_scale": float(speed_scale)
        * float(getattr(stage.options, "move_multiplier", 1.0) or 1.0),
        "ranged_enemies": bool(ranged_enemies),
        "enemy_windup": max(0.0, float(enemy_windup)),
    }
    # ---- 费用三项（`sim.py:474-481`）----
    out["cost_init"] = float(getattr(stage.options, "initial_cost", 0.0) or 0.0)
    out["cost_max"] = float(getattr(stage.options, "max_cost", 99.0) or 99.0)
    cost_time = float(getattr(stage.options, "cost_increase_time", 1.0) or 1.0)
    #: `cbuff_cost_recovery.scale`（四星档常见 2 = 回复速度翻倍）。它改的是
    #: **每点费用几秒**，所以要**除**。
    scale = cost_recovery_scale(stage, environment_difficulty)
    if scale and scale != 1.0:
        cost_time = cost_time / scale
    out["cost_time"] = cost_time
    # ---- 生命点（`sim.py:483-489`）----
    life = int(getattr(stage.options, "max_life_point", 1) or 1)
    lp = global_lifepoint(stage, environment_difficulty)
    if lp is not None:
        life = int(lp)
    out["life"] = life
    return out

File: rios-sim/stageenv.go
package main

import (
	"encoding/json"
	"fmt"
	"math"
	"os"
)

// stageenv.go：构建规格要用的**关卡静态 8 项**（丙阶段四·第十六批）。
//
// ## 为什么从这 8 项开刀
//
// `build_spec`（`simgo/spec.py:1161-1291`，131 行）是整个「Go 自己构造规格」
// 的入口，1291 行的 `spec.py` 里它要凑齐 **19 个顶层键**。一次搬完不现实，
// 所以按**依赖**切：哪些键不依赖 sim 状态、不依赖干员、不依赖机制？
//
// 这 8 项就是答案——它们只吃**关卡文件**（`options` ＋ `runes`），
// 而 Go 的 `StageOptions` 早就在读了，只差 `runes` 没留。
//
// 权威是 `ak_tactic/frontend/stage_env.py:68-102` 的 `stage_env`，
// 加上它的两个下游 `stage_mul.py` 的 `global_lifepoint` / `cost_recovery_scale`
// 与 `blackboard.py` 的三个取值原语。
//
// ## 两处必须照抄的「0 与 None 一视同仁」
//
// 原版到处写 `or`（`or 1.0` / `or 99.0` / `or 0.0`），于是一个**显式的 0**
// 会被换成缺省值。改成 `is None` 判断会在 `max_cost: 0` 这类取值上分叉，
// 而现有计划覆盖不到——所以下面每一处 `or` 都照抄成「等于零就取缺省」。
//
// ## 两条只改「每点费用几秒」与「生命点」的 rune
//
// * `cbuff_cost_recovery.scale`（四星档常见 2）——它改的是**每点费用几秒**，
//   所以要**除**。乘除弄反不会有任何判据报警，只会让费用回得快或慢。
// * `global_lifepoint.value`（八关 EX 的四星档都改成 1，而关卡文件自己写的
//   是 3、普通与四星**两份都是 3**）——不接这条，四星档凭空多两条命。
//
// ★ 而这两条**只在非 NORMAL 难度下**才与普通档不同，当前流水线恒为 NORMAL
// ⇒ 那是**沉默区**：重写出错没有任何判据会响。所以这里也是「原样搬」。

// BlackboardEntry 是 rune 黑板上的一条。数值住 `value`、字符串住 `valueStr`
// ——`bb_number` 只认前者（本项目记录在案的一次真错：只读 `valueStr` 会得出
// 「五个参数全是 None」，进而误判「环境系统不在 gamedata 里」）。
type BlackboardEntry struct {
	Key      string          `json:"key"`
	Value    json.RawMessage `json:"value"`
	ValueStr string          `json:"value_str,omitempty"`
}

// Rune 是一条关卡 rune。
type Rune struct {
	Key            string            `json:"key"`
	DifficultyMask *string           `json:"difficulty_mask"`
	Blackboard     []BlackboardEntry `json:"blackboard"`
}

func parseRunes(raw json.RawMessage) ([]Rune, error) {
	if len(raw) == 0 {
		return nil, nil
	}
	var arr []struct {
		Key            string  `json:"key"`
		DifficultyMask *string `json:"difficultyMask"`
		Blackboard     []struct {
			Key      string          `json:"key"`
			Value    json.RawMessage `json:"value"`
			ValueStr string          `json:"valueStr"`
		} `json:"blackboard"`
	}
	if err := json.Unmarshal(raw, &arr); err != nil {
		return nil, fmt.Errorf("runes 不是数组：%v", err)
	}
	out := make([]Rune, 0, len(arr))
	for _, r := range arr {
		rn := Rune{Key: r.Key, DifficultyMask: r.DifficultyMask}
		for _, b := range r.Blackboard {
			rn.Blackboard = append(rn.Blackboard, BlackboardEntry{
				Key: b.Key, Value: b.Value, ValueStr: b.ValueStr})
		}
		out = append(out, rn)
	}
	return out, nil
}

// maskApplies 复刻 `blackboard.py:42-47`。
//
// `ALL` 必须认——`act31side_08` 用的就是它，不认则整条环境系统静默消失。
func maskApplies(mask *string, difficulty string) bool {
	if mask == nil {
		return true
	}
	m := *mask
	return m == "ALL" || m == "" || m == difficulty
}

// findRune 复刻 `blackboard.py:50-63`。
//
// 遍历**全部**、命中就覆盖 ⇒ 同键同难度有多条时取**最后一条**。
// 「禁止取第一条」的来由：`act31side_ex08` 的 `env_system_new` 有 NORMAL 与
// FOUR_STAR 两条，黑板数值相同但 `init_pollut_value` 不同（`1,1:0` vs
// `4,4:100`）——取错了不报错，只是污染点画在了别处。
func findRune(runes []Rune, key, difficulty string) *Rune {
	var hit *Rune
	for i := range runes {
		if runes[i].Key == key && maskApplies(runes[i].DifficultyMask, difficulty) {
			hit = &runes[i]
		}
	}
	return hit
}

// bbNumber 复刻 `blackboard.py:66-78`：取**第一条**同键且值是数的条目。
func bbNumber(entries []BlackboardEntry, key string) *float64 {
	for _, e := range entries {
		if e.Key != key {
			continue
		}
		if len(e.Value) == 0 {
			return nil
		}
		var f float64
		if err := json.Unmarshal(e.Value, &f); err != nil {
			return nil
		}
		return &f
	}
	return nil
}

// globalLifepoint 复刻 `stage_mul.py:222-231`。老键名 `gbuff_lifepoint` 同样认。
func globalLifepoint(runes []Rune, difficulty string) *int {
	r := findRune(runes, "global_lifepoint", difficulty)
	if r == nil {
		r = findRune(runes, "gbuff_lifepoint", difficulty)
	}
	if r == nil {
		return nil
	}
	v := bbNumber(r.Blackboard, "value")
	if v == nil {
		return nil
	}
	n := int(*v)
	return &n
}

// costRecoveryScale 复刻 `stage_mul.py:234-246`：没有就是 1.0。
//
// ⚠ 末尾那句原版是 `float(v) if v else 1.0`——**取到的 0 也当「没有」**。
func costRecoveryScale(runes []Rune, difficulty string) float64 {
	r := findRune(runes, "cbuff_cost_recovery", difficulty)
	if r == nil {
		return 1.0
	}
	v := bbNumber(r.Blackboard, "scale")
	if v == nil || *v == 0 {
		return 1.0
	}
	return *v
}

// StageEnvOut 就是规格里那 8 个字段。
type StageEnvOut struct {
	FPS           int     `json:"fps"`
	SpeedScale    float64 `json:"speed_scale"`
	RangedEnemies bool    `json:"ranged_enemies"`
	EnemyWindup   float64 `json:"enemy_windup"`
	CostInit      float64 `json:"cost_init"`
	CostMax       float64 `json:"cost_max"`
	CostTime      float64 `json:"cost_time"`
	Life          int     `json:"life"`
}

// 后四个是 `BattleSimulator.__init__`（`sim.py:324-330`）的构造参数缺省值。
// 不传 switches 那条常见路径得到的就是这几个数。
const (
	stageEnvFPS           = 30
	stageEnvSpeedScale    = 1.0
	stageEnvRangedEnemies = true
	stageEnvEnemyWindup   = 0.5
)

// StageEnv 复刻 `stage_env.py:68-102`。
func StageEnv(opts StageOptions, runes []Rune, difficulty string) StageEnvOut {
	//: 每一处 `or` 都照抄成「等于零就取缺省」——显式 0 与缺省在这个口径下同义。
	moveMul := opts.MoveMultiplier
	if moveMul == 0 {
		moveMul = 1.0
	}
	costInit := opts.InitialCost //: 原版 `or 0.0`，取缺省也是 0，恒等
	costMax := opts.MaxCost
	if costMax == 0 {
		costMax = 99.0
	}
	costTime := opts.CostIncreaseTime
	if costTime == 0 {
		costTime = 1.0
	}
	life := opts.MaxLifePoint
	if life == 0 {
		life = 1
	}

	out := StageEnvOut{
		FPS:           stageEnvFPS,
		SpeedScale:    stageEnvSpeedScale * moveMul,
		RangedEnemies: stageEnvRangedEnemies,
		EnemyWindup:   math.Max(0.0, stageEnvEnemyWindup),
		CostInit:      costInit,
		CostMax:       costMax,
		CostTime:      costTime,
		Life:          life,
	}
	//: ⚠ `if scale and scale != 1.0` ——倍率为 0 时**不除**（那是「没有」）。
	if s := costRecoveryScale(runes, difficulty); s != 0 && s != 1.0 {
		out.CostTime = costTime / s
	}
	if lp := globalLifepoint(runes, difficulty); lp != nil {
		out.Life = *lp
	}
	return out
}

// LoadStageEnv 取一关的静态 8 项。
func LoadStageEnv(level, difficulty string) (StageEnvOut, error) {
	st, err := LoadStage(level)
	if err != nil {
		return StageEnvOut{}, err
	}
	if difficulty == "" {
		difficulty = "NORMAL"
	}
	return StageEnv(st.Options, st.Runes, difficulty), nil
}

// LoadStageEnvFile 同一个口径，但吃**一个关卡 JSON 文件路径**。
//
// 为什么要这条入口：判据要拿**合成关卡**把「显式 0 与缺省同义」那几条钉住
// （`max_cost: 0` → 99.0 之类），而合成关卡不在关卡索引里、`LoadStage` 找不到。
// 只解析 `options` 与 `runes` 两项，不走整份 `Stage` 的构造。
func LoadStageEnvFile(path, difficulty string) (StageEnvOut, error) {
	blob, err := os.ReadFile(path)
	if err != nil {
		return StageEnvOut{}, fmt.Errorf("关卡读不出来：%v", err)
	}
	var raw map[string]json.RawMessage
	if err := json.Unmarshal(blob, &raw); err != nil {
		return StageEnvOut{}, fmt.Errorf("关卡不是合法 JSON：%v", err)
	}
	opts, err := parseOptions(raw["options"])
	if err != nil {
		return StageEnvOut{}, fmt.Errorf("options：%w", err)
	}
	runes, err := parseRunes(raw["runes"])
	if err != nil {
		return StageEnvOut{}, fmt.Errorf("runes：%w", err)
	}
	if difficulty == "" {
		difficulty = "NORMAL"
	}
	return StageEnv(opts, runes, difficulty), nil
}
